swarm/network/stream/delivery.go

```diff
@@ -165,7 +82,6 @@ func (d *Delivery) handleRetrieveRequestMsg(ctx context.Context, sp *Peer, req *
 	go func() {
 		select {
 		case <-ctx.Done():
-		case <-streamer.quit:
```

Comment:
streamer.quit is canceling context in this goroutine on shutdown. If that is removed, then this goroutine is not doing anything as calling cancel on context that is done has no effect.

Reply:
Any suggestions how to fix this? Because of `RETRIEVE_REQUEST` stream, we no longer have `s` and we no longer have `streamer.

Reply:
I think that Delivery should have a Close method to terminate its goroutines.

Reply:
@janos I agree with you, but I think this is a feature we can add later. If I understand correctly, you want to have a clean node process which you will be able to start/stop in simulations and which handles the lifecycle of its go-routine correctly.

In a long-running Swarm process right now, I'd say that it doesn't matter if the go-routines are terminated when Delivery exits, because the main thread will exit.

I'd rather we focus on the rest of the fixes and bugs that we have already uncovered and solve this later - namely get the rest of the changes from `simplify-fetchers` in, and then address this.

We can focus on getting the gist of the simplified fetchers refactor in and address the subscription bug, so that we can release v0.4 or we can focus on every detail that is not handled correctly, and get there slower. I think we should do the former, but this is up for discussion :)

Reply:
I think that this is not a large feature, and that such things we should not leave unsolved. It is very important to correctly handle resources. If we skip on this types of resource handling, we will create more technical dept.

Reply:
done

Reply:
Thanks Anton.